File: backend/src/main/java/com/example/smart/lighting/scenes/with_natural/language/entity/Scene.java
package com.example.smart.lighting.scenes.with_natural.language.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.type.SqlTypes;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

/**
 * Entity representing a lighting scene (preset configuration).
 * Scenes define a snapshot of light settings that can be applied instantly.
 *
 * @author Smart Lighting Team
 * @version 1.0
 */
@Entity
@Table(name = "scenes", schema = "smartlighting")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Scene {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false, length = 100)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(length = 50)
    @Builder.Default
    private String icon = "💡";

    /**
     * Scene settings in JSON format.
     * Can be:
     * - All lights: {"target": "all", "brightness": 80, "rgb": [255,200,150], "color_temp": 3000}
     * - Specific lights: {"lights": [{"led_index": 0, "brightness": 80, ...}, ...]}
     */
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "settings_json", columnDefinition = "jsonb")
    @Builder.Default
    private Map<String, Object> settingsJson = new HashMap<>();

    /**
     * Preset scenes are system-defined and cannot be deleted by users.
     */
    @Column(name = "is_preset")
    @Builder.Default
    private Boolean isPreset = false;

    @Column(name = "is_active")
    @Builder.Default
    private Boolean isActive = true;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by")
    private User createdBy;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}

File: mobile/app/src/main/java/com/smartlighting/mobile/data/model/Scene.kt
package com.smartlighting.mobile.data.model

import com.google.gson.annotations.SerializedName

data class Scene(
    @SerializedName("id")
    val id: String? = null,
    @SerializedName("name")
    val name: String,
    @SerializedName("description")
    val description: String? = null,
    @SerializedName("icon")
    val icon: String? = null,
    @SerializedName("settings")
    val settings: SceneSettings? = null,
    @SerializedName("actions")
    val actions: List<SceneAction>? = null,
    @SerializedName("isPreset")
    val isPreset: Boolean? = false,
    @SerializedName("isGlobal")
    val isGlobal: Boolean? = false,
    @SerializedName("ownerId")
    val ownerId: String? = null,
    @SerializedName("ownerName")
    val ownerName: String? = null,
    @SerializedName("createdAt")
    val createdAt: String? = null,
    @SerializedName("updatedAt")
    val updatedAt: String? = null
)

data class SceneSettings(
    @SerializedName("brightness")
    val brightness: Int? = null,
    @SerializedName("rgb")
    val rgb: List<Int>? = null,
    @SerializedName("color_temp")
    val colorTemp: Int? = null,
    @SerializedName("target")
    val target: String? = null
)

data class SceneAction(
    @SerializedName("deviceId")
    val deviceId: String? = null,
    @SerializedName("on")
    val on: Boolean? = null,
    @SerializedName("brightness")
    val brightness: Int? = null,
    @SerializedName("colorTemp")
    val colorTemp: Int? = null,
    @SerializedName("rgbColor")
    val rgbColor: String? = null
)

